cheatsheets/gleam-for-typescript-users.md

```diff
@@ -0,0 +1,827 @@
+allows mixed types in the collection.
+
+```gleam
+let my_tuple = #("username", "password", 10)
+let #(_, pwd, _) = my_tuple
+io.print(pwd) // "password"
+// Direct index access
+io.print(my_tuple.0) // "username"
+```
+
+
+### Destructuring Arrays and Tuples
+
+#### TypeScript
+
+TypeScript supports destructuring arrays with `const` or `let` statements.
+
+```ts 
+type NumberArray = number[];
+type NumberPair = [number, number]; // tuple
+
+const [a, b]: NumberArray = [1, 2]; // the type annotation here is not required
+console.log(a); // 1
+console.log(b); // 2
+
+const [c]: NumberPair = [3, 4]; // the type annotation here is not required
+console.log(c); // 3
+
+const [d, ...e] = [5, 6, 7, 8];
+console.log(d); // 5
+console.log(e); // [6, 7, 8]
+```
+
+#### Gleam
+
+Similar syntax is supported in Gleam for tuples. However, variadic destructuring is not.
+
+```gleam
+let #(a, b) = #(1, 2)
+
+io.debug(a) // 1
+io.debug(b) // 2
+
+let #(c, _) = #(3, 4)
+io.debug(c) // 3
+// _ ignores the 4
+```
+
+For lists, the `list.split` method may be used.
+
+```gleam
+import gleam/io
+import gleam/list
+
+pub fn main() {
+  let ints = [11, 22, 33]
+
+  let #(a, rest) = list.split(ints, 1)
+
+  io.debug(a) // [11]
+  io.debug(rest) // [22, 33]
+}
+```
+
+### Objects, Records, and Maps
+
+The data structures in TypeScript and Gleam don't quite align one-to-one. TypeScript
+has objects (which can be typed as a record) and maps. Gleam has dictionaries.
+
+#### TypeScript
+
+```ts
+const obj = {
+  a: 1,
+  b: 'two',
+  c: {
+    neat: true,
+  }
+}
+
+obj.a; // 1
+obj['a']; // 1
+obj.c.neat; // true;
+```
+
+There are several ways to types objects in TypeScript. They provide a Record type,
+however, it is not a true record. Though, there is [a proposal to add true records
+and tuples](https://github.com/tc39/proposal-record-tuple) to JavaScript.
+
+```ts
+const obj1: { a: number } = { a: 1 };
+const obj2: { [key: string]: number } = { a: 1};
+const obj3: Record<string, number> = { a: 1 };
+
+{a: 1} === {a: 1} // false
+{a: 1} == {a: 1} // false
+```
+
+Maps in TypeScript are a class that are similar to objects, but separate.
+
+```ts
+const myMap = new Map<string, number>;
+
+myMap.set('a', 1);
+myMap.has('a'); // true
+myMap.get('a'); // 1
+```
+
+#### Gleam
+
+A dict, in Gleam, is a collection of keys and values which other languages may call a hashmap or table.
+
+```gleam
+import gleam/io
+import gleam/dict
+
+pub fn main() {
+  let scores = dict.from_list([#("Lucy", 13), #("Drew", 15)])
+  io.debug(scores) // dict.from_list([#("Drew", 15), #("Lucy", 13)])
+
+  let scores =
+    scores
+    |> dict.insert("Bushra", 16)
+    |> dict.insert("Darius", 14)
+    |> dict.delete("Drew")
+  io.debug(scores) // dict.from_list([#("Darius", 14), #("Bushra", 16), #("Lucy", 13)])
+}
+```
+
+### Sets
+
+A set is a collection of unique members. Both TypeScript and Gleam support sets.
+
+#### TypeScript
+
+A set in TypeScript can a set can have multiple types.
```

Comment:
Not quite- it has a single value type but TypeScript has subtyping. This would be true of everything in this guide so let's not bring it up specifically for sets.

The main difference is that in Gleam these data structures are immutable